File: day1/main.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

struct Input {
  vector<int> a, b;

  size_t size() const { return a.size(); }
};

void part1(Input input) {
  vector<int> &a = input.a;
  vector<int> &b = input.b;

  sort(a.begin(), a.end());
  sort(b.begin(), b.end());
  long long res = 0;
  for (int i = 0; i < input.size(); i++) {
    res += abs(a[i] - b[i]);
  }
  cout << res << endl;
}

void part2(Input input) {
  std::unordered_map<int, int> freq;
  for (const auto &x : input.b) {
    freq[x]++;
  }

  long long similarity = 0;
  for (const auto &x : input.a) {
    similarity += 1ll * x * freq[x];
  }

  cout << similarity << endl;
}

int main() {
  freopen("input.txt", "r", stdin);
  Input input;
  for (int a, b; cin >> a >> b;) {
    input.a.push_back(a);
    input.b.push_back(b);
  }
  part1(input);
  part2(input);
  return 0;
}

File: day1/main.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "sort"
    "strconv"
    "strings"
)

type Input struct {
    A []int
    B []int
}

func (i *Input) Len() int {
    return len(i.A)
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

func part1(input *Input) int {
    sort.Ints(input.A)
    sort.Ints(input.B)
    sum := 0
    for i := 0; i < input.Len(); i++ {
        sum += abs(input.A[i] - input.B[i])
    }
    return sum
}

func part2(input *Input) int {
    freqB := make(map[int]int)
    for _, b := range input.B {
        freqB[b]++
    }

    sum := 0
    for _, a := range input.A {
        sum += a * freqB[a]
    }
    return sum
}

func main() {
    file, _ := os.Open("input.txt")
    defer file.Close()

    var a, b []int
    scanner := bufio.NewScanner(file)

    for scanner.Scan() {
        line := scanner.Text()
        if line == "" {
            break
        }
        parts := strings.Fields(line)
        if x, err := strconv.Atoi(parts[0]); err == nil {
            a = append(a, x)
        }
        if x, err := strconv.Atoi(parts[1]); err == nil {
            b = append(b, x)
        }
    }

    input := &Input{a, b}

    fmt.Println(part1(input))
    fmt.Println(part2(input))
}
